client/trino-jdbc/src/test/java/io/trino/jdbc/TestJdbcPreparedStatement.java

```diff
@@ -1051,130 +1069,237 @@ public void testConvertTimestamp()
+    @Test(dataProvider = "legacyPreparedStatementProvider")
+    public void testInvalidConversions(boolean useLegacyPreparedStatements)
             throws SQLException
     {
-        assertBind((ps, i) -> ps.setObject(i, String.class)).isInvalid("Unsupported object type: java.lang.Class");
-        assertBind((ps, i) -> ps.setObject(i, String.class, Types.BIGINT)).isInvalid("Cannot convert instance of java.lang.Class to SQL type " + Types.BIGINT);
-        assertBind((ps, i) -> ps.setObject(i, "abc", Types.SMALLINT)).isInvalid("Cannot convert instance of java.lang.String to SQL type " + Types.SMALLINT);
+        assertBind((ps, i) -> ps.setObject(i, String.class), useLegacyPreparedStatements).isInvalid("Unsupported object type: java.lang.Class");
+        assertBind((ps, i) -> ps.setObject(i, String.class, Types.BIGINT), useLegacyPreparedStatements)
+                .isInvalid("Cannot convert instance of java.lang.Class to SQL type " + Types.BIGINT);
+        assertBind((ps, i) -> ps.setObject(i, "abc", Types.SMALLINT), useLegacyPreparedStatements)
+                .isInvalid("Cannot convert instance of java.lang.String to SQL type " + Types.SMALLINT);
+    }
+
+    @Test
+    public void testLegacyPreparedStatementsTrue()
+            throws Exception
+    {
+        testLegacyPreparedStatementsSetting(true,
+                "EXECUTE %statement% USING %values%");
+    }
+
+    @Test
+    public void testLegacyPreparedStatementsFalse()
+            throws Exception
+    {
+        testLegacyPreparedStatementsSetting(false,
+                "EXECUTE IMMEDIATE '%query%' USING %values%");
     }
 
-    private BindAssertion assertBind(Binder binder)
+    private BindAssertion assertBind(Binder binder, boolean useLegacyPreparedStatements)
     {
-        return new BindAssertion(this::createConnection, binder);
+        return new BindAssertion(() -> this.createConnection(useLegacyPreparedStatements), binder);
     }
 
-    private Connection createConnection()
+    private Connection createConnection(boolean useLegacyPreparedStatements)
             throws SQLException
     {
-        String url = format("jdbc:trino://%s", server.getAddress());
+        String url = format("jdbc:trino://%s?legacyPreparedStatements=" + useLegacyPreparedStatements, server.getAddress());
         return DriverManager.getConnection(url, "test", null);
     }
 
-    private Connection createConnection(String catalog, String schema)
+    private Connection createConnection(String catalog, String schema, boolean useLegacyPreparedStatements)
             throws SQLException
     {
-        String url = format("jdbc:trino://%s/%s/%s", server.getAddress(), catalog, schema);
+        String url = format("jdbc:trino://%s/%s/%s?legacyPreparedStatements=" + useLegacyPreparedStatements, server.getAddress(), catalog, schema);
         return DriverManager.getConnection(url, "test", null);
     }
 
+    private void checkSQLExecuted(Connection connection, String expectedSql)
```

Comment:
nit: move helpers AFTER the method which uses them

Reply:
Done